WIP async

ianb opened the pull request:
This is a start at addressing #21 

For some reason only some of the tests appear to me after I reload tests.html a few times, I don't understand quite what is happening.

This doesn't attempt to change index.js and the UI, it only attempts to address the tests currently.  There are failing tests, I think mostly around serializing access to streams between multiple tests, and in scope. Scope I suspect is quite broken; as it should be, I didn't touch it.  I really don't know how to handle scope.  Maybe trying to send it through to routines explicitly as a `this` argument?  The way it works currently I think just can't work in an async/concurrent environment.

Also BYE and STOP are totally unaddressed, and OUTPUT is kind of fuzzy.  I'm doing them as special ways of rejecting a promise now, which maybe just means at the top level there needs to be better catching of those rejections – but I don't quite understand what the "top" level is.  I think in the case of OUTPUT I actually found the appropriate place.  Is there nothing like `FOREVER [IF cond [BREAK]]` ?

There's a couple other FIXMEs of stuff I knowingly didn't do.

inexorabletash commented:
Each test case uses a separate interpreter instance so they should be isolated.

The tests assume synchronous behavior though, which may be why they're flaky for you. That'll have to change to being promise-based, which will require reworking the use of QUnit (https://api.qunitjs.com/async/) and the helpers.

ianb commented:
I setup the tests so they are async, though I believe the tests become somewhat interleaved in the process.  For instance, there's a test for `MAKE "t 10 :t` and later one that does `MaKe "t 5 :t` – that last set leaks into the first test so the result is 5.  (Since the scope relies on the call stack I'm not surprised it's broken.)

The problem I am having with the tests is that I get 41 tests run in the Parser section, and no other tests.   Here's what the tests look like to me: http://pageshot.dev.mozaws.net/mTG3Lb0YRafaKgBD/file – it could be some test setup problem I created, but there's nothing in the console.

inexorabletash commented on logo.js:
```diff
@@ -76,6 +76,14 @@ function LogoInterpreter(turtle, stream, savehook)
+  }
+
+  function makePromise(value) {
+    return isPromise(value) ? value : Promise.resolve(value);
```
You can replace makePromise with just `Promise.resolve(value)`

```js
var p = new Promise(function(){});
Promise.resolve(p) === p; // true
```

inexorabletash commented on tests.js:
```diff
@@ -71,10 +90,22 @@ QUnit.module("Logo Unit Tests", {
-      this.stream.clear();
-      t.equal(actual, expected, expression);
+      var result = this.interpreter.run(expression, {returnResult: true});
+      if (! (result && result.then)) {
```
Unnecessary - you can always just write `result = Promise.resolve(result)` which will be a no-op if already a Promise.

inexorabletash commented on tests.js:
```diff
@@ -86,13 +117,35 @@ QUnit.module("Logo Unit Tests", {
     this.assert_predicate = function(expression, predicate) {
-      t.ok(predicate(this.interpreter.run(expression, {returnResult: true})), expression);
+      var result = this.interpreter.run(expression, {returnResult: true});
+      if (result && result.then) {
```
This would be much simpler/easier to read if always async. So wrap in a Promise, or preferably just make interpreter.run return a Promise

ianb commented:
With c5918e0145404c0cf471eef706dd9de2ceac1887 the tests all seem to pass.  It does it mostly be removing all "real" concurrency, so everything is serialized, including the tests.  This is appropriate because we don't have any primitives for doing concurrent work (such as an event handler or something like setTimeout or setInterval).  So while the scope and stream bugs are not displaying themselves, they still exist.  (Well, maybe streams *should* be a jumble of all the concurrent actors.)

inexorabletash commented:
Thanks, I'll try and take a look, time permitting. I've left a few inline comments on style where the Promise using code can be simplified.

inexorabletash commented on logo.js:
```diff
@@ -2443,11 +2622,11 @@ function LogoInterpreter(turtle, stream, savehook)
 
   def("stop", function() {
-    throw new Output();
+    return Promise.reject({special: "stop"});
```
Why can't this simply be `return Promise.reject(new Output());` ?

But more importantly, if the procedure invoker does the right thing, this could still be `throw new Output();`

```js
// assume proc and inputs
var p = new Promise(function(resolve) {
  resolve(proc.apply(self, inputs));
});
```
* If proc returns a non-Promise, p will be a Promise resolved to that value
* If proc returns a Promise, p will be a Promise that resolves to the same value
* If proc throws, p will be a rejected Promise

inexorabletash commented on logo.js:
```diff
@@ -2379,32 +2523,67 @@ function LogoInterpreter(turtle, stream, savehook)
+    return new Promise(function (resolve, reject) {
+      var stop = false;
+      var i = 1;
+      function runLoop() {
```
ISTM guts of this should be as simple as `return self.execute(statement).then(runLoop);`

inexorabletash commented on tests.js:
```diff
@@ -60,39 +60,79 @@ QUnit.module("Logo Unit Tests", {
+        done();
+      }, function (failure) {
+        t.strictEqual(failure, expected, expression);
+        done();
```
Mentioned elsewhere, but you can move the redundant `done()` calls into a trailing `.then()` which behaves as a `finally` for promise chains.

inexorabletash commented on tests.js:
```diff
@@ -60,39 +60,79 @@ QUnit.module("Logo Unit Tests", {
     };
 
     this.assert_error = function(expression, expected) {
       try {
```
Seems out of place? Shouldn't be synchronous any more.

inexorabletash commented on logo.js:
```diff
@@ -546,6 +552,32 @@ function LogoInterpreter(turtle, stream, savehook)
 
+  function defer(func) {
+    var input = Array.prototype.slice.call(arguments, 1);
+    return function () {
```
I think this can be simplified to:

```js
return function() {
  return Promise.all(input).then(function(args) { return func.apply(null, args); });
};
```

ianb replied:
Promise.all() would evaluate the arguments out of order

inexorabletash replied:
Ah. :) Then let's factor out a helper function like Promise.all() but enforces the order.

inexorabletash replied:
Like this I think...

```js
function resolveAllInOrder(values) {
  return new Promise(function(resolve, reject) {
    var results = [];
    function next() {
      if (!values.length) {
        resolve(results);
      } else {
        Promise.resolve(values.shift)
          .then(function(result) { results.push(result); next(); })
          .catch(function(reason) { reject(reason); });
      }
    }();
  }
}
```

ianb replied:
Each item is a function that returns a promise sometimes (concrete expressions like `1` don't return promises), which makes it yet a bit more complicated, and doesn't feel very reusable.

inexorabletash commented on logo.js:
```diff
@@ -76,6 +76,10 @@ function LogoInterpreter(turtle, stream, savehook)
     return atom === match;
   }
 
+  function isPromise(value) {
```
In ECMAScript parlance, the duck-type is called a "thenable" and the existence of `then` is sufficient to identify it. Even in the ECMAScript spec itself when e.g. passing to `Promise.resolve()`

So maybe `isThenable`

ianb replied:
Yeah, I've seen the term, and yet it seems terribly obtuse doesn't it?

inexorabletash replied:
Yeah. But I think we shouldn't need this function anyway, be async everywhere, and Promise.resolve() to cast anything that might not already be a promise. 

> On Jan 31, 2016, at 8:25 PM, Ian Bicking <notifications@github.com> wrote:
> 
> In logo.js:
> 
> > @@ -76,6 +76,10 @@ function LogoInterpreter(turtle, stream, savehook)
> >      return atom === match;
> >    }
> >  
> > +  function isPromise(value) {
> Yeah, I've seen the term, and yet it seems terribly obtuse doesn't it?
> 
> —
> Reply to this email directly or view it on GitHub.
>

inexorabletash commented on logo.js:
```diff
@@ -710,7 +742,29 @@ function LogoInterpreter(turtle, stream, savehook)
-        return procedure.apply(null, args.map(function(a) { return a(); }));
+        var evaluatedArgs = [];
+        var index = 0;
+        return new Promise(function (resolve, reject) {
```
This looks like another case for `Promise.all()` ?

inexorabletash commented on logo.js:
```diff
@@ -818,21 +872,37 @@ function LogoInterpreter(turtle, stream, savehook)
+      function runLoop() {
+        while (statements.length) {
+          var result = evaluateExpression(statements);
+          if (isPromise(result)) {
```
else? Do we have any expressions that are not returning promises at this point? Internal error?

inexorabletash commented on logo.js:
```diff
@@ -841,19 +911,78 @@ function LogoInterpreter(turtle, stream, savehook)
+      self._pendingPromise = promise;
+      return promise;
     } catch (e) {
+      // FIXME: doesn't work with promises
```
Can you verify/explain? It "should" (in that "computers always make sense" sort of way...)

inexorabletash commented on logo.js:
```diff
@@ -841,19 +911,78 @@ function LogoInterpreter(turtle, stream, savehook)
+  // Helpers to ensure calls to run() are serialized:
+  //----------------------------------------------------------------------
+
+  // When there's an outstanding call to .run() (whose promise hasn't resolved
```
Just to confirm - the idea here is that if the promise returned by run() hasn't settled (e.g. it's waiting on a newfangled async `READWORD` prompt, or a long-running turtle operation, etc), and another run() comes in, we enqueue it then run it later?

Can I challenge that use case and say, for simplicity, that run() throws/rejects at that point, and its up to the caller to prevent that - e.g. the UI is disabled until the first `run` settles? In the UI we could turn Run into Stop which would interrupt execution in the case of an infinite loop. (Not handled today, I don't think)

ianb replied:
Yeah... it makes sense for the tests, but probably the tests could be fixed better by avoiding the reuse of interpreters.  For interactive work it doesn't really make sense.

inexorabletash replied:
We could make the assertions implicitly chain within a test - e.g. each one pushes onto a queue which is run sequentially when the test body returns.

inexorabletash commented on logo.js:
```diff
@@ -1600,10 +1738,12 @@ function LogoInterpreter(turtle, stream, savehook)
   def("true", function() { return 1; });
   def("false", function() { return 0; });
 
+// FIXME: needs to handle promises
```
Hrm... seems like the caller should handle deferring until the inputs are settled, so these procedures never see promises...

inexorabletash replied:
I guess I'd only expect special forms to ever see promises. Maybe higher-order procs - I haven't thought it all the way through yet.

ianb replied:
Because of `noeval` AND is a special form, it's evaluating its arguments lazily

inexorabletash replied:
D'oh! :) Yes, that makes perfect sense then. Don't mind me.

inexorabletash replied:
So we should probably throw (internal error) if a proc which is a special form returns a non-thenable.

inexorabletash commented on index.js:
```diff
@@ -360,8 +367,18 @@ function initInput() {
   $('#clear').addEventListener('click', clear);
 
+  function setRun(canRun) {
+    if (canRun) {
```
A better approach is to set/clear a style on an ancestor (e.g. `document.body.classList.add('running')` / `document.body.classList.remove('running')`) and then use CSS to control the visibility:

```css
#run, .running #stop { display: block; }
#stop, .running #run { display: none; }
```

ianb commented:
The example that prints "Logo" to the screen performs crazy slow after a few repetitions, not sure what's going on there.  Probably some crazy promise chain being constructed, though other long repeats don't have that same problem.

inexorabletash commented on index.css:
```diff
@@ -272,7 +272,7 @@ body {
 .multi  #toggle #expand-label { display: none; }
 
-#run, #clear {
+#run, #stop, #clear {
```
We should probably make this a `.button` class so we can stop appending IDs.

inexorabletash commented on logo.js:
```diff
@@ -76,6 +76,26 @@ function LogoInterpreter(turtle, stream, savehook)
+    return value && value.then && value.catch;
+  }
+
+  function promiseFinally(promise, finalBlock) {
```
I think this can be simplified to:

```js
function promiseFinally(promise, finalBlock) {
  return promise.then(finalBlock, finalBlock).then(function() { return promise; });
}
```

ianb replied:
I'm not sure what should happen when finalBlock rejects... should everything reject (maybe covering up a previous rejection), or not...?  Semantically a bit vague to me.

inexorabletash commented on logo.js:
```diff
@@ -818,40 +889,117 @@ function LogoInterpreter(turtle, stream, savehook)
-      // Parse it
-      var atoms = parse(string);
+    // Parse it
+    var atoms = parse(string);
```
Does state get cleaned up if `parse()` fails, now that there's no try/catch around it?

inexorabletash commented on logo.js:
```diff
@@ -1601,13 +1740,37 @@ function LogoInterpreter(turtle, stream, savehook)
 
+  function _checker(args, shouldStop) {
+    if (! args.length) {
+      return Promise.resolve(true);
```
Might be an existing bug, but what should `(AND)` and `(OR)` (i.e. called with 0 args) evaluate to?

inexorabletash commented on logo.js:
```diff
@@ -1673,23 +1837,31 @@ function LogoInterpreter(turtle, stream, savehook)
+          return result;
+        },
+        function (err) {
+          turtle.fillpath(fillcolor);
```
Can we add a comment that this is to handle exceptions as loop/proc terminators? (I know, I wrote the try/finally originally, but surprised me when re-reading the code).

inexorabletash replied:
Oh, and this can use promiseFinally right?

inexorabletash commented on logo.js:
```diff
@@ -1673,23 +1837,31 @@ function LogoInterpreter(turtle, stream, savehook)
+          turtle.fillpath(fillcolor);
+          throw err;
+        });
     } finally {
```
We can remove the try/finally now, right?

inexorabletash commented:
In case it's not obvious, I'm deliriously happy that you're tackling this. :)

Sorry about all the nitpicks - hopefully this feels like collaboration not complaining!

ianb commented:
Sure thing ;)  If you want to jump in we could move it to a branch on this repo too.

With the last commit it gets real concurrency, which is where I think it gets fun.  Debugging features like PAUSE are also fairly doable, though the UI of the interpreter itself starts to feel a bit constrained.

ianb commented:
Some process instructions from MicroWorlds:

* [done?](https://en.wikipedia.org/wiki/List_of_MicroWorlds_Logo_commands#done.3F)
* [everyone](https://en.wikipedia.org/wiki/List_of_MicroWorlds_Logo_commands#everyone)
* [forever](https://en.wikipedia.org/wiki/List_of_MicroWorlds_Logo_commands#forever) – it launches a process in MicroWorlds, I can see how it makes sense
* [get](https://en.wikipedia.org/wiki/List_of_MicroWorlds_Logo_commands#get) – gets an attribute from an object, and objects might be synonymous with processes.  Not sure how it relates to ASKSPAWN
* [glide](https://en.wikipedia.org/wiki/List_of_MicroWorlds_Logo_commands#glide) – slow version of FORWARD
* [launch](https://en.wikipedia.org/wiki/List_of_MicroWorlds_Logo_commands#launch) – like SPAWN.  Notably doesn't clone the turtle, so everyone gets to talk to the same turtle at once.
* [listen](https://en.wikipedia.org/wiki/List_of_MicroWorlds_Logo_commands#listen) – sets the default turtle that receives commands.  Processes and turtles aren't directly attached.
* [newbutton](https://en.wikipedia.org/wiki/List_of_MicroWorlds_Logo_commands#newbutton) – creates a button that I think is also attached to some instructions to be run when the button is clicked.  Label and instructions are identical...?
* [set](https://en.wikipedia.org/wiki/List_of_MicroWorlds_Logo_commands#set) – goes along with GET, separates objects and processes
* [setinstruction](https://en.wikipedia.org/wiki/List_of_MicroWorlds_Logo_commands#setinstruction) – attaches some instructions to an object, a little vague to me
* [stopall](https://en.wikipedia.org/wiki/List_of_MicroWorlds_Logo_commands#stopall) – stops everything in the world
* [stopme](https://en.wikipedia.org/wiki/List_of_MicroWorlds_Logo_commands#stopme) – stops the current process.  Same as BYE should do now.
* [talkto](https://en.wikipedia.org/wiki/List_of_MicroWorlds_Logo_commands#talkto) – seems the same as listen?
* [textwho](https://en.wikipedia.org/wiki/List_of_MicroWorlds_Logo_commands#textwho) – uses text boxes similar to turtles
* [turtlesown](https://en.wikipedia.org/wiki/List_of_MicroWorlds_Logo_commands#turtlesown) – globally reset properties on turtles?  I don't get this one.
* [who](https://en.wikipedia.org/wiki/List_of_MicroWorlds_Logo_commands#who) – gives name of the current turtle
* clickon, [clickoff](https://en.wikipedia.org/wiki/List_of_MicroWorlds_Logo_commands#clickoff) – these are mysterious to me, but referred to in some places around concurrency.

ianb commented:
clickon and clickoff is a pattern for giving objects their own autonomous code, where you can manually click them (e.g., to test) or just do `everyone [clickon]`.  A similar pattern to Hopscotch's Play button.

ianb commented:
I think I need to figure out a list of things to get this PR to "done", lest I get distracted before it gets merged.

There's a somewhat unintentional process model embedded in the patch, and I think it conflates a few things:

Implicitly global stuff:
- The list of processes
- The stream (console)
- Gensym index

There's a tree of processes, inheriting but allow local override of:
- Routines (maybe routines should all be global?)
- Property tables (I think property tables might be created implicitly, so this is kind of confusing)
- Scopes.  This creates a kind of closure around the scopes when the subprocess is created, that might be weird or hard to understand.  I haven't even thought through what it might mean in practice, except that the parent process's global scope gets inherited.

And locally each process has:
- A turtle (a confusing conflation of process and turtle)
- Random number stream (good?)
- Kind of has a process state (not well defined except for the ability to call .bye() from outside the process)
- A process name

A first run at a more sensible system:
- There's a new thing that contains processes, instead of reusing the parentmost process for this.
- It also contains turtles, the stream, gensym, the savehook
- I think it should also contain routines...?  This may be significant refactoring... or maybe not.  Maybe it would be as simple as taking the `self` argument, and replacing it with `this` and being careful how we get/invoke routines.  `this` would continue to be the process, just as `self` is now.

And for turtles:
- Turtles would be named, which seems a typical approach.
- `LISTEN` and `WHO` mostly control the turtles, plus something to create a new turtle.
- I don't think MicroWorld's special form `T1, FD 10` is necessary.  That form doesn't show up in their examples, for instance (they encourage other approaches that attach behavior to turtles).
- Maybe there should be an "object" system like in MicroWorlds with `GET` and `SET` and a global table of objects.  E.g., "T1 might be a turtle, "B1 might be a button, "P1 might be a property list, but all coming out of one namespace.  Though why not just be slightly modern and use :T1 etc?  I.e., accept objects into Logo as first-class values.  One reason why not: it can be nice for these objects to know their own name, might make it slightly more comprehensible.  And those names aren't dynamically scoped; it's "T1 everywhere and until it is changed.  Because these "objects" are actually concrete things sitting on the screen, it makes some sense that they exist outside of a variable scope.
- An approach some Logos take to these objects is to create routines for each object.  So T1 becomes a routine that returns the appropriate object, and it's global.  There might be a couple other accompanying routines using naming rules.
- Of course we can build up routines on top of this that make normal patterns easier.
- The "active turtle" could be a variable on the process, or maybe a standard variable name (so that `MAKE "ACTIVETURTLE "T1` is equivalent to `LISTEN "T1`).  I don't know if dynamic scoping helps or hurts.  There is type checking that the variable doesn't account for, like `MAKE "ACTIVETURTLE "doesnotexist`

Another approach is to try to land this without multiple processes at all.  Then there's no real concurrency, but it is possible for primitives to be async (like FD).  There's nothing stopping a non-concurrent Logo from still supporting objects and multiple turtles, but that keeps the issue separate.

inexorabletash commented:
I'd prefer to land it w/o multiple processes first.

inexorabletash commented:
Also, one refactor I'd like to make before adding a significant new number of library procedures would be to factor the library out into a separate module. That means no longer relying on the library procs being in a closure over the interpreter, and e.g. invoking procs explicitly with "this" as something meaningful (I think you mention this above).

ianb commented:
I addressed most of the comments I think.  Haven't put in any new tests though.  Removed the multiprocess stuff (moved it to [wip-concurrency](https://github.com/ianb/jslogo/tree/wip-concurrency) for now).

inexorabletash commented:
Cool. I'll try and do. Proper review pass and test things locally ASAP (probably a day or two). I'll aim to merge then apply any tweaks, if possible. 

> On Jan 31, 2016, at 8:38 PM, Ian Bicking <notifications@github.com> wrote:
> 
> I addressed most of the comments I think. Haven't put in any new tests though. Removed the multiprocess stuff (moved it to wip-concurrency for now).
> 
> —
> Reply to this email directly or view it on GitHub.
>

inexorabletash commented:
Bug: `cs` is not resetting the turtle's angle

inexorabletash commented:
Thoughts on the default speed? With the current speed, a lot of the turtle demos are glacially slow. Maybe 50 ? Also, until we can animate the turtle while drawing, perhaps we should make it "fast" by default?

inexorabletash commented on logo.js:
```diff
@@ -818,42 +889,118 @@ function LogoInterpreter(turtle, stream, savehook)
+      if (self.turtle) {
+        self.turtle.end();
+      }
+      if (err && (err.special == "bye" || err.special == "stop")) {
```
Where is `"stop"` coming from?

inexorabletash commented:
I've verified locally that the `{special: "...", ...}` form can be restored to using `Bye` and `Output`. Diff:

```diff
diff --git a/logo.js b/logo.js
index f293151..45add74 100644
--- a/logo.js
+++ b/logo.js
@@ -894,7 +894,7 @@ function LogoInterpreter(turtle, stream, savehook)
       function runLoop() {
         if (self.forceBye) {
           self.forceBye = false;
-          reject({special: "bye"});
+          reject(new Bye);
           return;
         }
         while (statements.length) {
@@ -944,8 +944,10 @@ function LogoInterpreter(turtle, stream, savehook)
       if (self.turtle) {
         self.turtle.end();
       }
-      if (err && (err.special == "bye" || err.special == "stop")) {
-        return undefined;
+      if (err instanceof Bye)
+        return;
+      if (err && err.special == "stop") {
+        return;
       }
       throw err;
     });
@@ -1150,9 +1152,8 @@ function LogoInterpreter(turtle, stream, savehook)
       }
       self.scopes.push(scope);
       return promiseFinally(self.execute(block).then(null, function (err) {
-        if (err && err.special == "output") {
-          return err.value;
-        }
+        if (err instanceof Output)
+          return err.output;
         throw err;
       }), function () {
         self.scopes.pop();
@@ -2623,11 +2624,11 @@ function LogoInterpreter(turtle, stream, savehook)
   });
 
   def("stop", function() {
-    return Promise.reject({special: "output", value: undefined});
+    throw new Output();
   });
 
   def(["output", "op"], function(atom) {
-    return Promise.reject({special: "output", value: atom});
+    throw new Output(atom);
   });
 
   // Not Supported: catch
@@ -2638,15 +2639,11 @@ function LogoInterpreter(turtle, stream, savehook)
   // Not Supported: wait
 
   def("bye", function() {
-    return Promise.reject({special: "bye"});
+    throw new Bye;
   });
 
   def(".maybeoutput", function(value) {
-    if (value !== undefined) {
-      return Promise.reject({special: "output", value: value});
-    } else {
-      return Promise.reject({special: "output"});
-    }
+    throw new Output(value);
   });
 
   // Not Supported: goto
``` 

Note that since procedure execution is wrapped in `Promise.resolve(...)` procs can just throw, which automagically turns into a rejection. Still not sure where `"stop"` came from - dead end?

inexorabletash commented:
Let me know when you're ready and I'll push this to a branch so we can both hack on it.

inexorabletash commented:
This enables most of the tests and fixes `AND` and `OR`. It explicitly disables the failing test blocks since most of them hang. We should work through the failures to get them all passing.

```diff
diff --git a/logo.js b/logo.js
index 45add74..f158550 100644
--- a/logo.js
+++ b/logo.js
@@ -1741,33 +1741,32 @@ function LogoInterpreter(turtle, stream, savehook)
 
   def("and", function(a, b) {
     var args = Array.from(arguments);
-    return checker(args, function (value) {return ! value;}, 1);
+    return _checker(args, function (value) {return !value;}, 1);
   }, {noeval: true});
 
   def("or", function(a, b) {
     var args = Array.from(arguments);
-    return checker(args, function (value) {return value;}, 0);
+    return _checker(args, function (value) {return value;}, 0);
   }, {noeval: true});
 
   function _checker(args, shouldStop, defaultValue) {
-    if (! args.length) {
+    if (!args.length)
       return Promise.resolve(defaultValue);
-    }
-    return new Promise(function (resolve, reject) {
-      function runLoop() {
-        if (args.length == 1) {
-          return resolve(args[0]());
+    return new Promise(function(resolve, reject) {
+      (function runLoop() {
+        var r = args.shift()();
+        if (!args.length) {
+          resolve(r);
+          return;
         }
-        var result = Promise.resolve(args.shift());
-        result.then(function (value) {
+        Promise.resolve(r).then(function(value) {
           if (shouldStop(value)) {
             resolve(value);
+            return;
           }
           runLoop();
-        }, function (err) {
-          reject(err);
         });
-      }
+      }());
     });
   }
 
@@ -2559,7 +2558,7 @@ function LogoInterpreter(turtle, stream, savehook)
         }
         self.repcount = i;
         i++;
-        result = self.execute(statements);
+        var result = self.execute(statements);
         result.then(runLoop, reject);
       }
       runLoop();
diff --git a/tests.html b/tests.html
index 51c86ec..343725e 100644
--- a/tests.html
+++ b/tests.html
@@ -4,8 +4,8 @@
 
 <script src="polyfill/polyfill.js"></script>
 
-<script src="https://code.jquery.com/qunit/qunit-1.20.0.js"></script>
-<link rel="stylesheet" href="https://code.jquery.com/qunit/qunit-1.20.0.css" type="text/css" media="screen" />
+<script src="https://code.jquery.com/qunit/qunit-1.21.0.js"></script>
+<link rel="stylesheet" href="https://code.jquery.com/qunit/qunit-1.21.0.css">
 
 <h1 id="qunit-header">Logo Unit Tests</h1>
 <h2 id="qunit-banner"></h2>
diff --git a/tests.js b/tests.js
index 56be711..fcf9481 100644
--- a/tests.js
+++ b/tests.js
@@ -124,9 +124,11 @@ QUnit.module("Logo Unit Tests", {
           done();
         }, function (ex) {
           t.push(ex.message === expected, ex.message, expected, 'Expected error from: ' + expression);
+          done();
         });
       } catch (ex) {
         t.push(ex.message === expected, ex.message, expected, 'Expected error from: ' + expression);
+        done();
       }
     };
   }
@@ -138,6 +140,7 @@ QUnit.test("Parser", function(t) {
   //
 
   this.assert_equals('"test', 'test');
+
   this.assert_equals('1', 1);
   this.assert_equals('[ a b c ]', ["a", "b", "c"]);
   this.assert_equals('[ 1 2 3 ]', ["1", "2", "3"]);
@@ -496,7 +499,7 @@ QUnit.test("Data Structure Primitives", function(t) {
   this.assert_equals('standout "whatever', 'whatever');
 });
 
-QUnit.test("Communication", function(t) {
+if (false) QUnit.test("Communication", function(t) {
   t.expect(22);
 
   // 3.1 Transmitters
@@ -757,7 +760,7 @@ QUnit.test("Logical Operations", function(t) {
   this.assert_stream('or 0 (type "yup)', 'yup');
 });
 
-QUnit.test("Graphics", function(t) {
+if (false) QUnit.test("Graphics", function(t) {
   t.expect(69);
 
   // NOTE: test canvas is 300,300 (so -150...150 coordinates before hitting)
@@ -1051,7 +1054,7 @@ QUnit.test("Workspace Management", function(t) {
   // 7.7 Workspace Control
 });
 
-QUnit.test("Control Structures", function(t) {
+if (false) QUnit.test("Control Structures", function(t) {
   t.expect(58);
   //
   // 8.1 Control
@@ -1160,7 +1163,7 @@ QUnit.test("Control Structures", function(t) {
   // TODO: Structures, lists of lists
 });
 
-QUnit.test("Error Messages", function(t) {
+if (false) QUnit.test("Error Messages", function(t) {
   this.assert_error("to foo end show foo", "No output from procedure");
   this.assert_error("[ 1 2", "Expected ']'");
   this.assert_error("{ 1 2", "Expected '}'");
@@ -1225,7 +1228,7 @@ QUnit.test("Error Messages", function(t) {
   this.assert_error('make "a { 1 2 3 }@1.5', "Don't know what to do with 0.5");
 });
 
-QUnit.test("Regression Tests", function(t) {
+if (false) QUnit.test("Regression Tests", function(t) {
   this.assert_equals('make "x 0  repeat 3 [ for [ i 1 4 ] [ make "x :x + 1 ] ]  :x', 12);
   this.assert_equals('make "x 0  for [i 0 100 :i + 1] [make "x :x + :i]  :x', 120);
   this.assert_error("fd 100 50 rt 90", "Don't know what to do with 50");
@@ -1259,7 +1262,7 @@ QUnit.test("Regression Tests", function(t) {
   this.assert_equals('make "a { 1 }  make "b :a  setitem 1 :a 2  item 1 :b', 2);
 });
 
-QUnit.test("API Tests", function(t) {
+if (false) QUnit.test("API Tests", function(t) {
   // LogoInterpeter#copydef(newname, oldname)
   this.assert_error('yup', "Don't know how to YUP");
   this.assert_error('nope', "Don't know how to NOPE");
```

inexorabletash commented:
I suppose I could have made those PRs against your branch. *sigh*

ianb commented:
I don't get any test failures myself, which I think confirms my suspicion that lots of the tests aren't running.  But what's up with that?  I'm just loading test.html and letting it run.

inexorabletash commented:
The tests aren't finishing due to bugs in the code execution. Eg for AND and OR the wrong function name was used, and the arg fund wasn't being called. 

I haven't debugged the rest. 

> On Feb 2, 2016, at 6:55 AM, Ian Bicking <notifications@github.com> wrote:
> 
> I don't get any test failures myself, which I think confirms my suspicion that lots of the tests aren't running. But what's up with that? I'm just loading test.html and letting it run.
> 
> —
> Reply to this email directly or view it on GitHub.
>

inexorabletash commented:
Also, yes, the async tests are interfering with each other. I'm not sure QUnit's mechanism can handle multiple async assertions, or if it's bugs in logo yet. 

> On Feb 2, 2016, at 6:55 AM, Ian Bicking <notifications@github.com> wrote:
> 
> I don't get any test failures myself, which I think confirms my suspicion that lots of the tests aren't running. But what's up with that? I'm just loading test.html and letting it run.
> 
> —
> Reply to this email directly or view it on GitHub.
>

inexorabletash commented:
The run queue wasn't working properly; I rewrote it:

```js
  function noop() {}
  var lastRun = Promise.resolve();

  self.run = function(string, options) {
    var promise = lastRun.then(function() {
      options = Object(options);
      // Parse it                                                                                                                               
      var atoms = parse(string);

      if (self.turtle)
        self.turtle.begin();

      // And execute it!                                                                                                                        
      var p = self.execute(atoms, options);


      // TODO: use promiseFinally() here                                                                                                        
      p.catch(noop).then(function() {
        if (self.turtle)
          self.turtle.end();
      });

      return p.catch(function(err) {
        if (err instanceof Bye)
          return;
        throw err;
      });
    });
    lastRun = promise.catch(noop);
    return promise;
  };
```

With this, `self._setPendingPromise`, `self._unsetPendingPromise` and `self._queueCall` can be removed, and all of the Error Message tests pass. QUnit's async mechanism works fine, yay.

inexorabletash commented:
Some Communication tests fail because the `inputbuffer` is synchronously modified by the test but asynchronously used. And something is funky in the `assert_prompt` logic.

In both cases, we probably need the assert functions to return promises themselves so they can be chained.

inexorabletash commented:
Fix for RUNRESULT:

```diff
   def("runresult", function(statements) {
     statements = reparse(lexpr(statements));
-    var result = self.execute(statements, {returnResult: true});
-    if (isPromise(result)) {
-      return result.then(function (r) {
-        return [r];
+    return self.execute(statements, {returnResult: true})
+      .then(function(result) {
+        if (result !== undefined)
+          return [result];
+        else
+          return [];
       });
-    } else if (result !== undefined) {
-      return [result];
-    } else {
-      return [];
-    }
   });
```

inexorabletash commented:
Here's a rewritten FOR that passes the Control Structure tests. do.while and do.until still need fixing:

```js

  def("for", function(control, statements) {
    control = reparse(lexpr(control));
    statements = reparse(lexpr(statements));

    function sign(x) { return x < 0 ? -1 : x > 0 ? 1 : 0; }

    var varname = sexpr(control.shift());
    var start, limit, step, current;

    return Promise.resolve(evaluateExpression(control))
      .then(function(r) {
        current = start = aexpr(r);
        return evaluateExpression(control);
      })
      .then(function(r) {
        limit = aexpr(r);
      })
      .then(function() {
        return new Promise(function(resolve, reject) {
          function runLoop() {
            if (sign(current - limit) === sign(step)) {
              resolve();
              return;
            }
            setvar(varname, current);
            self.execute(statements)
              .then(function() {
                return (control.length) ?
                  evaluateExpression(control.slice()) : sign(limit - start);
              })
              .then(function(result) {
                step = aexpr(result);
                current += step;
                runLoop();
              });
          }
          runLoop();
        });
      });
  });
```

inexorabletash commented:
Simplified DO.WHILE and WHILE:

```js
  def("do.while", function(block, tf) {
    block = checkevalblock(block);
    return new Promise(function(resolve, reject) {
      (function loop() {
        self.execute(block)
          .then(tf)
          .then(function(cond) {
            if (!cond) {
              resolve();
              return;
            }
            loop();
          })
          .catch(reject);
      }());
    });
  }, {noeval: true});

  def("while", function(tf, block) {
    block = checkevalblock(block);
    return new Promise(function(resolve, reject) {
      (function loop() {
        Promise.resolve(tf())
          .then(function(cond) {
            if (!cond) {
              resolve();
              return;
            }
            self.execute(block)
              .then(loop);
          })
          .catch(reject);
      }());
    });
  }, {noeval: true});
```

And working DO.UNTIL and UNTIL:

```js
 function notpromise(tf) {
    return Promise.resolve(tf()).then(function(r) { return !r; });
  }

  def("do.until", function(block, tf) {
    var nottf = function () { return notpromise(tf); };
    return self.routines.get("do.while")(block, nottf);
  }, {noeval: true});

  def("until", function(tf, block) {
    var nottf = function () { return notpromise(tf); };
    return self.routines.get("while")(nottf, block);
  }, {noeval: true});

```

inexorabletash commented:
CASE is fine already, no work needed.

Here's COND

```js
  def("cond", function(clauses) {
    clauses = lexpr(clauses);

    return new Promise(function(resolve, reject) {
      (function loop() {
        if (!clauses.length) {
          resolve();
          return;
        }
        var clause = lexpr(clauses.shift());
        var first = clause.shift();
        if (isKeyword(first, 'ELSE')) {
          resolve(evaluateExpression(clause));
          return;
        }
        evaluateExpression(reparse(lexpr(first)))
          .then(function(result) {
            if (result) {
              resolve(evaluateExpression(clause));
              return;
            }
            loop();
          })
          .catch(reject);
      }());
    });
  });
```

inexorabletash commented:
... and that gets all of the Control Structure tests passing.

The Regression Tests also all pass except for

> `this.assert_error("fd 100 50 rt 90", "Don't know what to do with 50");`

... but I'm done for the night.
